File: app/src/main/java/morawski/rafal/barcodescanner/scanner/ScannerBoxOverlay.java
package morawski.rafal.barcodescanner.scanner;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.view.View;

import com.google.android.gms.common.images.Size;

public class ScannerBoxOverlay extends View {
    private static final String TAG = "ScannerBoxOverlay";

    private int mLeft;
    private int mTop;
    private int mRectWidth = 200;
    private int mRectHeight = 200;
    private float mStrokeWidth = 10.0f;
    private float mCornerRadius = 20.0f;
    private int mCrosshairColor;

    private Paint mEraser;
    private Paint mCrosshairPaint;

    public ScannerBoxOverlay(Context context) {
        super(context);
        init();
    }

    private void init() {
        mEraser = new Paint();
        mEraser.setAntiAlias(true);
        mEraser.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.CLEAR));
        mEraser.setStrokeWidth(mStrokeWidth);

        mCrosshairColor = Color.WHITE;

        mCrosshairPaint = new Paint();
        mCrosshairPaint.setColor(mCrosshairColor);
        mCrosshairPaint.setStrokeWidth(mStrokeWidth);
        mCrosshairPaint.setStyle(Paint.Style.STROKE);
    }

    public void setScanBoxSizePx(Size size) {
        mRectWidth = size.getWidth();
        mRectHeight = size.getHeight();
        invalidate();
    }

    public Size getBoxSize() {
        return new Size(mRectWidth, mRectHeight);
    }

    public void setCrosshairColor(int color) {
        mCrosshairColor = color;
        invalidate();
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        int width = l + r;
        int height = t + b;

        mLeft = (width - mRectWidth) / 2;
        mTop = (height - mRectHeight) / 2;
    }

    @Override
    protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight) {
        super.onSizeChanged(width, height, oldWidth, oldHeight);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        // Draw transparent rect;
        canvas.drawRect(mLeft, mTop, mLeft + mRectWidth, mTop + mRectHeight, mEraser);

        // Draw crosshair
        drawCrosshair(canvas, mCrosshairColor);
    }

    private void drawCrosshair(Canvas canvas, int color) {
        int left = (int) (mLeft - mStrokeWidth / 2);
        int top = (int) (mTop - mStrokeWidth / 2);
        int right = (int) (mLeft + mRectWidth + mStrokeWidth / 2);
        int bottom = (int) (mTop + mRectHeight + mStrokeWidth / 2);

        mCrosshairPaint.setColor(color);
        canvas.drawRoundRect(left, top, right, bottom, mCornerRadius, mCornerRadius, mCrosshairPaint);

        int widthOffset = (int) (mRectWidth * 0.3 / 2.0);
        int heightOffset = (int) (mRectHeight * 0.3 / 2.0);
        int offset = Math.min(widthOffset, heightOffset);

        mEraser.setStrokeWidth(mCrosshairPaint.getStrokeWidth());
        canvas.drawLine(left + offset, top, right - offset, top, mEraser);
        canvas.drawLine(left + offset, bottom, right - offset, bottom, mEraser);
        canvas.drawLine(left, top + offset, left, bottom - offset, mEraser);
        canvas.drawLine(right, top + offset, right, bottom - offset, mEraser);
    }
}

File: app/src/main/java/morawski/rafal/barcodescanner/scanner/ScannerBoxOverlay.kt
package morawski.rafal.barcodescanner.scanner

import android.content.Context
import android.graphics.*
import android.view.View
import com.google.android.gms.common.images.Size
import kotlin.math.min

class ScannerBoxOverlay(context: Context) : View(context) {

    private var mLeft: Int = 0
    private var mTop: Int = 0
    private var mRectWidth: Int = 200
    private var mRectHeight: Int = 200
    private var mStrokeWidth: Float = 10.0f
    private var mCornerRadius: Float = 20.0f
    private var mViewfinderColor: Int = Color.WHITE

    private lateinit var mEraser: Paint
    private lateinit var mViewfinderPaint: Paint

    init {
        initialize()
    }

    private fun initialize() {
        mEraser = Paint().apply {
            isAntiAlias = true
            xfermode = PorterDuffXfermode(PorterDuff.Mode.CLEAR)
            strokeWidth = mStrokeWidth
        }

        mViewfinderColor = Color.WHITE

        mViewfinderPaint = Paint().apply {
            color = mViewfinderColor
            strokeWidth = mStrokeWidth
            style = Paint.Style.STROKE
        }
    }

    fun setBoxSizePx(size: Size) {
        mRectWidth = size.width
        mRectHeight = size.height
        invalidate()
    }

    fun getBoxSizePx() = Size(mRectWidth, mRectHeight)

    fun setViewfinderColor(color: Int) {
        mViewfinderColor = color
        invalidate()
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        val width = left + right
        val height = top + bottom

        mLeft = (width - mRectWidth) / 2
        mTop = (height - mRectHeight) / 2
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)

        val canv = canvas ?: return
        // Draw transparent rect
        canv.drawRect(mLeft.toFloat(), mTop.toFloat(), (mLeft + mRectWidth).toFloat(), (mTop + mRectHeight).toFloat(), mEraser)

        // Draw viewfinder
        drawViewfinder(canv, mViewfinderColor)
    }

    private fun drawViewfinder(canvas: Canvas, color: Int) {
        val left = mLeft - mStrokeWidth / 2
        val top = mTop - mStrokeWidth / 2
        val right = mLeft + mRectWidth + mStrokeWidth / 2
        val bottom = mTop + mRectHeight + mStrokeWidth / 2

        mViewfinderPaint.color = color
        canvas.drawRoundRect(left, top, right, bottom, mCornerRadius, mCornerRadius, mViewfinderPaint)

        val widthOffset = (mRectWidth * 0.3 / 2.0).toInt()
        val heightOffset = (mRectHeight * 0.3 / 2.0).toInt()
        val offset = min(widthOffset, heightOffset)

        mEraser.strokeWidth = mViewfinderPaint.strokeWidth
        canvas.drawLine(left + offset, top, right - offset, top, mEraser)
        canvas.drawLine(left + offset, bottom, right - offset, bottom, mEraser)
        canvas.drawLine(left, top + offset, left, bottom - offset, mEraser)
        canvas.drawLine(right, top + offset, right, bottom - offset, mEraser)
    }
}